modules/katamari/src/main/ts/ephox/katamari/api/Optional.ts

```diff
@@ -1,141 +1,291 @@
+   * `replacement` value if the provided `Optional` object does not contain a
+   * value.  Unlike `getOr`, in this method the `replacement` object is also
+   * `Optional` - meaning that this method will always return an `Optional`.
+   */
+  public or<U = T>(replacement: Optional<U>): Optional<T | U> {
+    return this.tag ? this : replacement;
+  }
+
+  /**
+   * Get the value out of the inside of the `Optional` object, using a default
+   * `replacement` value if the provided `Optional` object does not contain a
+   * value. Unlike `getOr`, in this method the `replacement` value is
+   * "thunked" - that is to say that you don't pass a value to `getOrThunk`, you
+   * pass a function which (if called) will **return** the `value` you want to
+   * use.
+   */
+  public getOrThunk<U = T>(thunk: () => U): T | U {
+    return this.tag ? this.value as T : thunk();
+  }
+
+  /**
+   * Get the value out of the inside of the `Optional` object, using a default
+   * `replacement` value if the provided Optional object does not contain a
+   * value.
+   *
+   * Unlike `or`, in this method the `replacement` value is "thunked" - that is
+   * to say that you don't pass a value to `getOrThunk`, you pass a function
+   * which (if called) will **return** the `value` you want to use.
+   *
+   * Unlike `getOrThunk`, in this method the `replacement` value is also
+   * `Optional`, meaning that this method will always return an `Optional`.
+   */
+  public orThunk<U = T>(thunk: () => Optional<U>): Optional<T | U> {
+    return this.tag ? this : thunk();
+  }
+
+  /**
+   * Get the value out of the inside of the `Optional` object, throwing an
+   * exception if the provided `Optional` object does not contain a value.
+   *
+   * WARNING:
+   * You should only be using this function if you know that the `Optional`
+   * object **is not** empty (otherwise you're throwing exceptions in production
+   * code, which is bad).
+   *
+   * In tests this is more acceptable.
+   *
+   * Prefer other methods to this, such as `.each`.
+   */
+  public getOrDie(message?: string): T {
+    if (!this.tag) {
+      throw new Error(message ?? 'Called getOrDie on None');
+    } else {
+      return this.value as T;
+    }
+  }
+
+  // --- Interop with null and undefined ---
+
+  /**
+   * Creates an `Optional` value from a nullable (or undefined-able) input.
+   * Null, or undefined, is converted to `None`, and anything else is converted
+   * to `Some`.
+   */
+  public static from<T>(this: void, value: T | null | undefined): Optional<NonNullable<T>> {
+    return Type.isNonNullable(value) ? Optional.some(value) : Optional.none();
+  }
+
+  /**
+   * Converts an `Optional` to a nullable type.
+   */
+  public getOrNull(): T | null {
+    return this.tag ? this.value as T : null;
+  }
+
+  /**
+   * Converts an `Optional` to an undefined-able type.
+   */
+  public getOrUndefined(): T | undefined {
+    return this.tag ? this.value as T : undefined;
```

Comment:
Could this be just `return this.value;`?

Reply:
It could, good catch thanks